Support getAutoGeneratedKeys when using executeBatch in readwritespli…

baijun44 opened the pull request:
…tting or single database (#24294)

Changes proposed in this pull request:
  -Support getAutoGeneratedKeys when using executeBatch in readwritesplitting or single database
  -Add unit test

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

strongduanmu commented on jdbc/core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/statement/ShardingSpherePreparedStatement.java:
```diff
@@ -653,7 +658,17 @@ public int[] executeBatch() throws SQLException {
-            return batchPreparedStatementExecutor.executeBatch(executionContext.getSqlStatementContext());
+            int[] results = batchPreparedStatementExecutor.executeBatch(executionContext.getSqlStatementContext());
+            if (statementOption.isReturnGeneratedKeys() && batchPreparedStatementExecutor.getStatements().size() == 1 && generatedValues.isEmpty()) {
+                ResultSet resultSet = batchPreparedStatementExecutor.getStatements().get(0).getGeneratedKeys();
```
Why only get the result generated by the first statement?

baijun44 replied:
I put a condition which requires statements size must equal 1. when using write db or single db, the statements size is 1. when using sharding mode. the statements size may greater than 1. In sharding mode, We almost use keygenerators .So I think just supporting single mode and rw mode is enough. I also test get all statement result in sharding mode,it can work too. but I hope reduce scope of influence.

strongduanmu commented on jdbc/core/src/test/java/org/apache/shardingsphere/driver/jdbc/core/statement/ReadwriteSplittingPreparedStatementTest.java:
```diff
@@ -118,4 +118,65 @@ public void assertGetGeneratedKeysWithPrimaryKeyIsNullInTransactional() throws S
     }
+    
+    @Test
+    public void assertGetAutoGeneratedKeysAfterExecuteBatch() throws SQLException {
```
Can you move these case to e2e test case? We need to test different scenarios such as sharding, read-write splitting, etc.

strongduanmu replied:
You can add test case in `cases/dml/dml-integration-insert.xml` file, and then execute `BatchDMLE2EIT`.

baijun44 replied:
Alright,I will try this week

strongduanmu commented:
Hi @baijun44, does this issue has any progress?

baijun44 commented:
> Hi @baijun44, does this pr has any progress?

I will update the code today.

baijun44 commented:
> Hi @baijun44, does this pr has any progress?

Hi,I try to add test case but I can't run it on my mac. Could you help me? I am very glad to recieve your guidance.Thank you!

strongduanmu commented:
> > Hi @baijun44, does this pr has any progress?
> 
> Hi,I try to add test case but I can't run it on my mac. Could you help me? I am very glad to recieve your guidance.Thank you!

Hi @baijun44, you can refer https://github.com/apache/shardingsphere/blob/master/.github/workflows/e2e.yml. If you want to run e2e on your local machine, you should build docker image first, and then modify `it-env.properties` in suite module.

TeslaCN commented:
I'm moving this to 5.3.3.

strongduanmu commented:
Hi @baijun44, does this issue has any progress?

baijun44 commented:
> Hi @baijun44, does this issue has any progress?

Yes, It's not easy to complete this work😂, there is no existing test cases for autoGeneratedKeys by database itself. I wrote a new test case and I found something wrong here.Looking forward to your reply🥺

boyjoy1127 commented:
Will this issue be completed before June 15 which is the final date of version 5.4.0?

baijun44 commented:
> Will this issue be completed before June 15 which is the final date of version 5.4.0?

maybe not, I think it will be completed in next verion.

liming0 commented:
> > Hi @baijun44, does this pr has any progress?
> 
> Hi,I try to add test case but I can't run it on my mac. Could you help me? I am very glad to recieve your guidance.Thank you!

I have a pull request that was unsuccessful during the CI phase, but there is a compilation issue with this project on my computer, which seems to be related to those. G4 files. I don't know how to solve it. Can you help me?

baijun44 commented:
> Hi @baijun44, does this issue has any progress?

Hi,@strongduanmu, finally I finished this issue, I refactored some codes and added test cases. Could you help me review these codes? Thanks!

baijun44 commented:
Hi, @strongduanmu , I noticed that a failed test in the workflow. So I re-ran on my pc. But success. It may not be caused by my PR. I rebased my PR onto the latest branch.Could you help re-run?

strongduanmu commented:
@baijun44 Can you try e2e on your local machine?

baijun44 commented:
> @baijun44 Can you try e2e on your local machine?

Hi, @strongduanmu ,I fixed some bugs,Please help me review.Thanks!

baijun44 commented:
> Hi @baijun44, e2e is failed, can you try to fix it?

Yes, I'm dealing with it. I am glad to see that most of the use cases have been passed. Thank you for your patience and help.
